File: matlab/figures/table07_canonical.m
function table07_canonical()
% table07_canonical  Regenerate the paper's Table VI (pooled statistics)
% under the CANONICAL single-state conventions of the revised manuscript:
%
%   * links, LOS/NLOS labels, and N1/U1 values are taken from the canonical
%     point-data xlsx (the "orig." columns of N3_142/U3_142/N3_7/U3_7 --
%     identical to the released supplementary tables);
%   * OLOS is relabeled NLOS for pooled modeling;
%   * CI path-loss fit with 1 m free-space reference, no intercept,
%     sigma_SF with ddof = 0; bootstrap 95% CFI over links (B = 2000);
%   * DS/ASA/ASD lognormal means exp(mu*ln10 + (sigma*ln10)^2/2) on log10
%     values (ddof = 0), full-width bootstrap CFIs;
%   * the declared 2 ns delay-resolution floor is EXECUTED for the DS fits
%     (resolution-limited links are excluded from DS distribution fits and
%     retained everywhere else).
%
% Output: <repo>/figures/matlab/table07_canonical.csv
% This is the authoritative regeneration path for the revised Table VI; the
% legacy table07_pooled_stats (hybrid CSV/xlsx loader) is retained only for
% provenance of the previous submission.

P = paths();
out_csv = fullfile(P.out_dir, 'table07_canonical.csv');
DS_FLOOR_NS = 2.0;
B_BOOT = 2000;
rng(0);

sets = {
    % institution, band label, band GHz, xlsx path
    'NYU', 'Sub-THz',  142.0, P.n3_142_xlsx;
    'USC', 'Sub-THz',  145.5, P.u3_142_xlsx;
    'NYU', '6.75 GHz', 6.75,  P.n3_7_xlsx;
    'USC', '6.75 GHz', 6.75,  P.u3_7_xlsx;
};

% -- Load all links ---------------------------------------------------------
L = table();
for i = 1:size(sets, 1)
    T = read_canonical(sets{i, 4});
    T.institution = repmat(string(sets{i, 1}), height(T), 1);
    T.band        = repmat(string(sets{i, 2}), height(T), 1);
    T.f_ghz       = repmat(sets{i, 3}, height(T), 1);
    L = [L; T]; %#ok<AGROW>
end
L.cls = L.loc_type;
L.cls(L.cls == "OLOS") = "NLOS";

fprintf('[table07_canonical] %d valid links loaded\n', height(L));

% -- Fit every (band, dataset, class) cell ----------------------------------
rows = {};
bands = ["Sub-THz", "6.75 GHz"];
dsets = {"NYU only", "USC only", "Pooled"};
for b = bands
    for d = 1:numel(dsets)
        for cls = ["LOS", "NLOS"]
            sel = L.band == b & L.cls == cls;
            if d < 3
                inst = extractBefore(string(dsets{d}), ' ');
                sel = sel & L.institution == inst;
            end
            g = L(sel, :);

            [ple, sig] = ci_fit(g.dist_m, g.pl, g.f_ghz);
            cfi = boot_ci_fit(g.dist_m, g.pl, g.f_ghz, B_BOOT);

            ds_v  = g.ds(~isnan(g.ds) & g.ds >= DS_FLOOR_NS);
            asa_v = g.asa(~isnan(g.asa) & g.asa > 0);
            asd_v = g.asd(~isnan(g.asd) & g.asd > 0);

            [ds_m, ds_w]   = logn_mean_ci(ds_v, B_BOOT);
            [asa_m, asa_w] = logn_mean_ci(asa_v, B_BOOT);
            [asd_m, asd_w] = logn_mean_ci(asd_v, B_BOOT);

            rows(end+1, :) = {char(b), dsets{d}, char(cls), height(g), ...
                round(ple, 2), round(sig, 2), round(cfi, 2), ...
                round(ds_m, 2), round(ds_w, 2), numel(ds_v), ...
                round(asa_m, 2), round(asa_w, 2), numel(asa_v), ...
                round(asd_m, 2), round(asd_w, 2), numel(asd_v)}; %#ok<AGROW>
        end
    end
end

T = cell2table(rows, 'VariableNames', {'Band', 'Dataset', 'LocType', ...
    'n_links', 'PLE', 'sigma_SF_dB', 'PLE_CFI_width', ...
    'DS_mean', 'DS_CFI_width', 'DS_n', ...
    'ASA_mean', 'ASA_CFI_width', 'ASA_n', ...
    'ASD_mean', 'ASD_CFI_width', 'ASD_n'});
writetable(T, out_csv);
fprintf('[table07_canonical] wrote %s\n', out_csv);
end

% ===========================================================================
function T = read_canonical(xlsx_path)
% Two-row-header canonical xlsx -> table(tx, rx, loc_type, dist_m, pl, ds,
% asa, asd) using the institution-original ("orig.") columns 8/11/14/17.
c = readcell(char(xlsx_path), 'Sheet', 'FinalTable');
tx = strings(0); rx = strings(0); loc = strings(0);
dist = []; pl = []; ds = []; asa = []; asd = [];
cur_tx = "";
for r = 4:size(c, 1)
    v_tx = cellstr_or_empty(c{r, 2});
    if strlength(v_tx) > 0, cur_tx = v_tx; end
    v_rx = cellstr_or_empty(c{r, 3});
    if strlength(v_rx) == 0, continue; end
    d  = num_or_nan(c{r, 5});
    p  = num_or_nan(c{r, 8});
    s  = num_or_nan(c{r, 11});
    aa = num_or_nan(c{r, 14});
    ad = num_or_nan(c{r, 17});
    if isnan(d) || (isnan(p) && isnan(s) && isnan(aa) && isnan(ad))
        continue;  % outage row
    end
    tx(end+1, 1) = cur_tx; rx(end+1, 1) = v_rx; %#ok<AGROW>
    loc(end+1, 1) = upper(cellstr_or_empty(c{r, 4})); %#ok<AGROW>
    dist(end+1, 1) = d; pl(end+1, 1) = p; ds(end+1, 1) = s; %#ok<AGROW>
    asa(end+1, 1) = aa; asd(end+1, 1) = ad; %#ok<AGROW>
end
T = table(tx, rx, loc, dist, pl, ds, asa, asd, 'VariableNames', ...
    {'tx', 'rx', 'loc_type', 'dist_m', 'pl', 'ds', 'asa', 'asd'});
end

function s = cellstr_or_empty(v)
if ismissing(v), s = ""; return; end
if isnumeric(v), s = string(v); return; end
s = strtrim(string(v));
end

function y = num_or_nan(v)
if isnumeric(v) && isscalar(v), y = double(v);
elseif ischar(v) || isstring(v), y = str2double(v);
else, y = NaN;
end
end

function [n, sig] = ci_fit(d, pl, f)
m = ~isnan(pl);
x = 10 .* log10(d(m));
y = pl(m) - fspl1m(f(m));
n = sum(x .* y) / sum(x .^ 2);
r = y - n .* x;
sig = sqrt(mean(r .^ 2) - mean(r) ^ 2 + mean(r) ^ 2); % population std
sig = std(r, 1);
end

function w = boot_ci_fit(d, pl, f, B)
m = ~isnan(pl);
d = d(m); pl = pl(m); f = f(m);
nl = numel(d);
vals = zeros(B, 1);
for b = 1:B
    idx = randi(nl, nl, 1);
    vals(b) = ci_fit(d(idx), pl(idx), f(idx));
end
q = quantile(vals, [0.025, 0.975]);
w = q(2) - q(1);
end

function y = fspl1m(f_ghz)
y = 20 .* log10(4 * pi .* f_ghz .* 1e9 ./ 299792458);
end

function [mu_ln, w] = logn_mean_ci(x, B)
lx = log10(x);
ln10 = log(10);
mu_ln = exp(mean(lx) * ln10 + 0.5 * (std(lx, 1) * ln10) ^ 2);
nl = numel(x);
vals = zeros(B, 1);
for b = 1:B
    i = randi(nl, nl, 1);
    v = lx(i);
    vals(b) = exp(mean(v) * ln10 + 0.5 * (std(v, 1) * ln10) ^ 2);
end
q = quantile(vals, [0.025, 0.975]);
w = q(2) - q(1);
end

File: python/scripts/table07_canonical.py
"""table07_canonical.py -- bit-exact regeneration of the revised paper's
Table VII (pooled statistics) from the canonical point-data xlsx.

Conventions (single released state, revised manuscript):
  - links/labels/values from the "orig." columns of N3_142/U3_142/N3_7/U3_7
    (identical to the released supplementary tables); OLOS -> NLOS;
  - CI path-loss fit (1 m free-space reference, no intercept, sigma ddof=0);
  - lognormal means exp(mu*ln10 + (sigma*ln10)^2/2) on log10 values (ddof=0);
  - bootstrap 95% CFIs, full width, B=2000, numpy default_rng(seed=0) per
    cell in fixed cell order (bit-exact regeneration of the published CFIs);
  - the declared 2 ns delay-resolution floor is EXECUTED for DS fits.

Usage:  python python/scripts/table07_canonical.py
Output: figures/python/table07_canonical.csv

The MATLAB counterpart (matlab/figures/table07_canonical.m) reproduces all
deterministic columns exactly; its bootstrap CFI widths differ within RNG
convention (<8%).
"""

from pathlib import Path

import numpy as np
import pandas as pd

REPO = Path(__file__).resolve().parents[2]
PD_DIR = REPO / "data" / "point_data"
OUT = REPO / "figures" / "python"
OUT.mkdir(parents=True, exist_ok=True)

C = 299_792_458.0
B_BOOT = 2000
SEED = 0
DS_FLOOR_NS = 2.0

# Order matters for bit-exact bootstrap CFIs: concat order fixes the link
# order inside pooled cells (matches the paper's generator).
FILES = [
    ("USC", "Sub-THz", 145.5, "U3_142_UMi.xlsx"),
    ("USC", "6.75 GHz", 6.75, "U3_7_UMi.xlsx"),
    ("NYU", "Sub-THz", 142.0, "N3_142_UMi.xlsx"),
    ("NYU", "6.75 GHz", 6.75, "N3_7_UMi.xlsx"),
]


def read_canonical(path):
    raw = pd.read_excel(path, header=None, sheet_name="FinalTable")
    rows = []
    cur_tx = ""
    for r in range(3, len(raw)):
        tx = raw.iat[r, 1]
        if isinstance(tx, str) and tx.strip():
            cur_tx = tx.strip()
        rx = raw.iat[r, 2]
        if not (isinstance(rx, str) and rx.strip()):
            continue
        vals = [pd.to_numeric(raw.iat[r, c], errors="coerce")
                for c in (4, 7, 10, 13, 16)]
        dist, pl, ds, asa, asd = vals
        if pd.isna(dist) or all(pd.isna(v) for v in (pl, ds, asa, asd)):
            continue  # outage row
        rows.append(dict(tx=cur_tx, rx=str(rx).strip(),
                         loc_type=str(raw.iat[r, 3]).strip().upper(),
                         dist_m=dist, pl=pl, ds=ds, asa=asa, asd=asd))
    return pd.DataFrame(rows)


def fspl_1m_db(f_ghz):
    return 20.0 * np.log10(4.0 * np.pi * np.asarray(f_ghz, float) * 1e9 / C)


def ci_fit(d, pl, f):
    x = 10.0 * np.log10(np.asarray(d, float))
    y = np.asarray(pl, float) - fspl_1m_db(f)
    n = float(np.sum(x * y) / np.sum(x * x))
    return n, float(np.std(y - n * x, ddof=0))


def ci_boot_width(d, pl, f, rng):
    d, pl, f = (np.asarray(a, float) for a in (d, pl, f))
    m = len(d)
    vals = np.empty(B_BOOT)
    for b in range(B_BOOT):
        i = rng.integers(0, m, m)
        vals[b], _ = ci_fit(d[i], pl[i], f[i])
    lo, hi = np.percentile(vals, [2.5, 97.5])
    return float(hi - lo)


def lognormal_mean(x):
    lx = np.log10(np.asarray(x, float))
    ln10 = np.log(10.0)
    return float(np.exp(lx.mean() * ln10 + 0.5 * (lx.std(ddof=0) * ln10) ** 2))


def lognormal_boot_width(x, rng):
    x = np.asarray(x, float)
    m = len(x)
    vals = np.empty(B_BOOT)
    for b in range(B_BOOT):
        vals[b] = lognormal_mean(x[rng.integers(0, m, m)])
    lo, hi = np.percentile(vals, [2.5, 97.5])
    return float(hi - lo)


def main() -> None:
    frames = []
    for inst, band, f_ghz, name in FILES:
        t = read_canonical(PD_DIR / name)
        t["institution"], t["band"], t["f_ghz"] = inst, band, f_ghz
        frames.append(t)
    df = pd.concat(frames, ignore_index=True)
    df["cls"] = df.loc_type.replace({"OLOS": "NLOS"})
    print(f"{len(df)} valid links loaded")

    rows = []
    for band in ["Sub-THz", "6.75 GHz"]:
        for dataset in ["NYU only", "USC only", "Pooled"]:
            for cls in ["LOS", "NLOS"]:
                sel = (df.band == band) & (df.cls == cls)
                if dataset != "Pooled":
                    sel &= df.institution == dataset.split()[0]
                g = df[sel]
                rng = np.random.default_rng(SEED)
                p = g.dropna(subset=["pl"])
                ple, sig = ci_fit(p.dist_m, p.pl, p.f_ghz)
                cfi = ci_boot_width(p.dist_m, p.pl, p.f_ghz, rng)
                row = dict(Band=band, Dataset=dataset, LocType=cls,
                           n_links=len(g), PLE=round(ple, 2),
                           sigma_SF_dB=round(sig, 2),
                           PLE_CFI_width=round(cfi, 2))
                for name_, col, floor in [("DS", "ds", DS_FLOOR_NS),
                                          ("ASA", "asa", 0.0),
                                          ("ASD", "asd", 0.0)]:
                    v = g[col].dropna()
                    v = v[v >= floor] if floor else v[v > 0]
                    rng2 = np.random.default_rng(SEED)
                    row[f"{name_}_mean"] = round(lognormal_mean(v), 2)
                    row[f"{name_}_CFI_width"] = round(
                        lognormal_boot_width(v, rng2), 2)
                    row[f"{name_}_n"] = len(v)
                rows.append(row)

    out = pd.DataFrame(rows)
    out.to_csv(OUT / "table07_canonical.csv", index=False)
    print(f"wrote {OUT / 'table07_canonical.csv'}")
    print(out.to_string(index=False))


if __name__ == "__main__":
    main()
